### ISSUE TITLE
Memory leak in `useSharedValue`

### ISSUE THREAD

**Issue by Lapz** [software-mansion/react-native-reanimated#5800]
### Description

The memory used by a `useSharedValue` is never released when the component is unmounted and is always held

### Steps to reproduce

Create a flatlist component that renders some subcomponents, the subcomponents all `useSharedValues`. Swipe through the list and you'll see that the amount of memory keeps on increasing and it's never released. Running the hermes profiler and you'll see that the `useSharedValue` comes up a lot with the number of allocations. I've attached a repo below where you can take a look, this seems to be similar to #5614

### Snack or a link to a repository

https://github.com/Lapz/reanimated-shared-values-leak

### Reanimated version

3.8.0

### React Native version

0.73.5

### Platforms

iOS

### JavaScript runtime

Hermes

### Workflow

Expo Go

### Architecture

Paper (Old Architecture)

### Build type

None

### Device

iOS simulator

### Acknowledgements

Yes

**Comment by MehediH** [comment]
I am also experiencing the same, @piaskowyk would be great if you or anyone from the team could kindly investigate. Thank you so much! 🙇🏽

**Comment by efstathiosntonas** [comment]
On the latest nightly (01 April '24) running the example on iOS 17.2 using Xcode 15.2:

1. Run the example
2. Tap RuntimeTestsExample row
3. Tap Run tests
4. Let it run for a while

On terminal:

```sh
leaks ReanimatedExample 
```

results in:

```sh
Analysis Tool Version:  iOS Simulator 17.2 (21C62)

Physical footprint:         430.7M
Physical footprint (peak):  461.8M
Idle exit:                  untracked
----

leaks Report Version: 4.0
Process 12646: 232495 nodes malloced for 357758 KB
Process 12646: 422 leaks for 9792 total leaked bytes.

    422 (9.56K) << TOTAL >>
      1 (1008 bytes) ROOT LEAK: 0x108973580 [1008]  length: 9  "REAModule"
      1 (1008 bytes) ROOT LEAK: 0x1089769b0 [1008]  length: 34  "reanimated::NativeReanimatedModule"
      1 (1008 bytes) ROOT LEAK: 0x12d49e610 [1008]  length: 9  "REAModule"
...
...
...
```

trying to trace the leak: ` leaks --traceTree=0x12d49e610 ReanimatedExample` results to:

```
Tracing: 0x12d49e610 [1008]  length: 9  "REAModule"

Found 0 roots referencing: 0x12d49e610 [1008]  length: 9  "REAModule"

This is a reverse reference tree showing how the specified block is referenced.
The top node is the specified block, the next level down are blocks that reference that, and so on.
The number at the start of each line is the number of paths-to-roots going through that node.
This reverse reference tree can be imported into Instruments using Import Sample Data to allow data mining.
```

Not quite sure why it fails to trace it.

**Comment by efstathiosntonas** [comment]
@tjzel @piaskowyk can you please investigate? Lot's of memory leak issues have been risen lately :/

In my lists the memory spikes at 1.5GB on an iPhone 15 Pro Max after 150-200 items, I don't want to imagine what happens with low end devices 😅

**Comment by efstathiosntonas** [comment]
my knowledge on C:  0

I asked ChatGPT about the `SingleInstanceChecker.h` under `Common/cpp/Tools` to suggest some improvements and it gave me back this:

<details>
  <summary>Click me</summary>
  
```c
#pragma once

#ifndef NDEBUG

#include <cxxabi.h>

#include <atomic>
#include <iostream>
#include <string>

#ifdef ANDROID
#include <android/log.h>
#endif

namespace reanimated {

// This is a class that counts how many instances of a different class there
// are. It is meant only to be used with classes that should only have one
// instance.

template <class T>
class SingleInstanceChecker {
 public:
  SingleInstanceChecker();
  ~SingleInstanceChecker();

 private:
  void assertWithMessage(bool condition, std::string message) {
    if (!condition) {
#ifdef ANDROID
      __android_log_print(
          ANDROID_LOG_WARN, "Reanimated", "%s", message.c_str());
#else
      std::cerr << "[Reanimated] " << message << std::endl;
#endif

#ifdef IS_REANIMATED_EXAMPLE_APP
      assert(false);
#endif
    }
  }

  // A static field will exist separately for every class template.
  // This automatically initializes the inline variable.
  inline static std::atomic<int> instanceCount_{0};
};

template <class T>
SingleInstanceChecker<T>::SingleInstanceChecker() {
  int status = 0;
  char* demangled_name = __cxxabiv1::__cxa_demangle(typeid(T).name(), nullptr, nullptr, &status);

  std::string className =
      (status == 0 && demangled_name != nullptr) ? demangled_name : typeid(T).name();

  assertWithMessage(
      instanceCount_.load() <= 1,
      "[Reanimated] More than one instance of " + className +
          " present. This may indicate a memory leak due to a retain cycle.");

  instanceCount_.fetch_add(1);

  free(demangled_name);
}

template <class T>
SingleInstanceChecker<T>::~SingleInstanceChecker() {
  instanceCount_.fetch_sub(1);
}

} // namespace reanimated

#endif // NDEBUG
```
</details>



The leaks are gone. I have no idea what I'm doing so I don't know if there are any side effects. Just thought it would be nice to share.

**Comment by efstathiosntonas** [comment]
In a Flashlist with posts I use my lib to cache the images: https://github.com/georstat/react-native-image-cache

The lib transitions the thumbnail image with the actual image using timing so it's a heavy user of `useAnimatedStyle`, `useSharedValue` and `withTiming`  when it comes to scrolling.

After 150-200 posts the RAM skyrockets to 1.5G as stated in a previous comment. The moment I switch the image component to `expo-image` the RAM stays at 500MB.

**Comment by MehediH** [comment]
@efstathiosntonas You are correct. We were using `react-native-fast-image` and switching to Expo Image solved the memory leak issue. Now when components are unmounted, they actually release the memory.

**Comment by efstathiosntonas** [comment]
@MehediH this is a common issue with fast-image (that's the main reason I created the lib mentioned in the previous comment). 

Problem is, the issue exists in the official ReanimatedExample  too.

**Comment by devym-37** [comment]
Could this issue be related to the app freezing issue on Android with the useSharedValue issue ?

**Comment by PEZO19** [comment]
@efstathiosntonas Can you please update how your fix [mentioned above](https://github.com/software-mansion/react-native-reanimated/issues/5800#issuecomment-2032631035) is working? Have you used it since in a live project too? If so, how was the "integration" part of your fix?

**Comment by efstathiosntonas** [comment]
@PEZO19 hi, haven’t used it in production since I have no clue if what chatgpt proposed has any side effects in other parts of reanimated, I just posted that for testing purposes or to give some possible hints to SW devs.

**Comment by szydlovsky** [comment]
Hey everyone, I confirmed that we are aiming to take a look at this issue. However, it's not on the top of our priorities for now. It will probably come up in July. Stay tuned!

**Comment by avivash** [comment]
hi @szydlovsky 👋🏼 just wondering if your team has had a chance to look into this yet?

**Comment by szydlovsky** [comment]
Hey @avivash we are currently researching issues related to performance - but not this exact topic yet. If anything changes I will let you know! And, of course, I will always encourage trying to find the culprit yourself 😄

**Comment by MehediH** [comment]
@szydlovsky hey! happy holidays - curious to see if there's an update here?

**Comment by piaskowyk** [comment]
Hey @MehediH 👋 I definitely agree that this is an important issue and we want to address it in the future, but currently, we're running out of team capacity to handle it right now. We need more time :/

**Comment by alexfoxy** [comment]
I am noticing an issue after upgrading my app to the new architecture. Images inside a component that use a sharedValue are not released from memory. Is there any way to force this? I'm just using the bare React Native "Image" component.

**Comment by alexfoxy** [comment]
@piaskowyk do you have any suggestions for a work around ?

**Comment by drminh2807** [comment]
It's also happened with Reanimated v4 without `useSharedValue`. RAM increased and was never released
```js
          backgroundColor,
          transitionProperty: 'backgroundColor',
          transitionDuration: 300,
```

**Comment by rahulnainwalttn** [comment]
+1 

Facing the same issue, RAM is continuously getting increased. @piaskowyk Any update on it?

**Comment by exzib** [comment]
Any update on this?

**Comment by tshmieldev** [comment]
@exzib Possibly fixed in https://github.com/software-mansion/react-native-reanimated/pull/9393
